cargo nightly does not resolve nuklear-sys

nuklear-backend-gfx 0.1.3 does not build on nightly, but does on stable. The reason seems to be that Cargo is not downloading and building nuklear-sys. Here's the output from a failed run:

```
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading gfx_window_glutin v0.12.0
 Downloading gfx_device_gl v0.11.2
 Downloading glutin v0.6.2
 Downloading nuklear-rust v0.1.5
 Downloading gfx v0.12.1
 Downloading gfx_core v0.4.0
 Downloading draw_state v0.6.0
 Downloading bitflags v0.6.0
 Downloading gfx_gl v0.3.1
 Downloading gl_generator v0.5.2
 Downloading khronos_api v1.0.0
 Downloading xml-rs v0.3.4
 Downloading libc v0.2.17
 Downloading shared_library v0.1.5
 Downloading x11-dl v2.9.0
 Downloading wayland-kbd v0.3.6
 Downloading osmesa-sys v0.1.2
 Downloading wayland-client v0.5.12
 Downloading wayland-window v0.2.3
 Downloading dlib v0.3.0
 Downloading libloading v0.2.4
 Downloading target_build_utils v0.1.2
 Downloading serde_json v0.8.3
 Downloading phf v0.7.17
 Downloading dtoa v0.2.2
 Downloading serde v0.8.14
 Downloading num-traits v0.1.36
 Downloading phf_shared v0.7.17
 Downloading phf_codegen v0.7.17
 Downloading phf_generator v0.7.17
 Downloading crossbeam v0.2.10
 Downloading wayland-sys v0.5.11
 Downloading wayland-scanner v0.5.11
   Compiling semver v0.1.20
   Compiling bitflags v0.6.0
   Compiling crossbeam v0.2.10
   Compiling khronos_api v1.0.0
   Compiling pkg-config v0.3.8
   Compiling phf_shared v0.7.17
   Compiling winapi v0.2.8
   Compiling bitflags v0.7.0
   Compiling winapi-build v0.1.1
   Compiling x11-dl v2.9.0
   Compiling phf v0.7.17
   Compiling dtoa v0.2.2
   Compiling kernel32-sys v0.2.2
   Compiling byteorder v0.5.3
   Compiling serde v0.8.14
   Compiling lazy_static v0.1.16
   Compiling num-traits v0.1.36
   Compiling log v0.3.6
   Compiling rustc_version v0.1.7
   Compiling draw_state v0.6.0
   Compiling itoa v0.1.1
   Compiling tempfile v2.1.4
   Compiling libc v0.2.17
   Compiling gfx_core v0.4.0
   Compiling lazy_static v0.2.1
   Compiling xml-rs v0.3.4
   Compiling shared_library v0.1.5
   Compiling nuklear-rust v0.1.5
error[E0463]: can't find crate for `nuklear_sys`
 --> /home/brian/.cargo/registry/src/github.com-1ecc6299db9ec823/nuklear-rust-0.1.5/src/lib.rs:5:1
  |
5 | extern crate nuklear_sys;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error

```

And here's what happened when I used stable cargo:

```
 Downloading nuklear-sys v1.156.0
 Downloading gcc v0.3.37
   Compiling lazy_static v0.2.1
   Compiling serde v0.8.14
   Compiling libc v0.2.17
   Compiling crossbeam v0.2.10
   Compiling winapi v0.2.8
   Compiling log v0.3.6
   Compiling num-traits v0.1.36
   Compiling shared_library v0.1.5
   Compiling itoa v0.1.1
   Compiling osmesa-sys v0.1.2
   Compiling khronos_api v1.0.0
   Compiling winapi-build v0.1.1
   Compiling byteorder v0.5.3
   Compiling kernel32-sys v0.2.2
   Compiling semver v0.1.20
   Compiling rand v0.3.14
   Compiling phf_shared v0.7.17
   Compiling rustc_version v0.1.7
   Compiling phf v0.7.17
   Compiling pkg-config v0.3.8
   Compiling lazy_static v0.1.16
   Compiling bitflags v0.6.0
   Compiling draw_state v0.6.0
   Compiling x11-dl v2.9.0
   Compiling phf_generator v0.7.17
   Compiling tempfile v2.1.4
   Compiling bitflags v0.7.0
   Compiling gfx_core v0.4.0
   Compiling xml-rs v0.3.4
   Compiling wayland-scanner v0.5.11
   Compiling gl_generator v0.5.2
   Compiling phf_codegen v0.7.17
   Compiling target_build_utils v0.1.2
   Compiling gfx_gl v0.3.1
...
```

cc @snuk182

--- a/src/cargo/core/manifest.rs
+++ b/src/cargo/core/manifest.rs
@@ -185,7 +185,8 @@ impl Encodable for Target {
 }
 
 impl Manifest {
-    pub fn new(summary: Summary, targets: Vec<Target>,
+    pub fn new(summary: Summary,
+               targets: Vec<Target>,
                exclude: Vec<String>,
                include: Vec<String>,
                links: Option<String>,

--- a/src/cargo/sources/registry/mod.rs
+++ b/src/cargo/sources/registry/mod.rs
@@ -359,7 +359,19 @@ impl<'cfg> Source for RegistrySource<'cfg> {
         }));
         let mut src = PathSource::new(&path, &self.source_id, self.config);
         try!(src.update());
-        src.download(package)
+        let pkg = try!(src.download(package));
+
+        // Unfortunately the index and the actual Cargo.toml in the index can
+        // differ due to historical Cargo bugs. To paper over these we trash the
+        // *summary* loaded from the Cargo.toml we just downloaded with the one
+        // we loaded from the index.
+        let summaries = try!(self.index.summaries(package.name()));
+        let summary = summaries.iter().map(|s| &s.0).find(|s| {
+            s.package_id() == package
+        }).expect("summary not found");
+        let mut manifest = pkg.manifest().clone();
+        manifest.set_summary(summary.clone());
+        Ok(Package::new(manifest, pkg.manifest_path()))
     }
 
     fn fingerprint(&self, pkg: &Package) -> CargoResult<String> {

--- a/tests/cargotest/support/registry.rs
+++ b/tests/cargotest/support/registry.rs
@@ -141,7 +141,7 @@ impl Package {
             map.insert("name".to_string(), dep.name.to_json());
             map.insert("req".to_string(), dep.vers.to_json());
             map.insert("features".to_string(), dep.features.to_json());
-            map.insert("default_features".to_string(), false.to_json());
+            map.insert("default_features".to_string(), true.to_json());
             map.insert("target".to_string(), dep.target.to_json());
             map.insert("optional".to_string(), false.to_json());
             map.insert("kind".to_string(), dep.kind.to_json());
@@ -211,7 +211,7 @@ impl Package {
         for dep in self.deps.iter() {
             let target = match dep.target {
                 None => String::new(),
-                Some(ref s) => format!("target.{}.", s),
+                Some(ref s) => format!("target.'{}'.", s),
             };
             let kind = match &dep.kind[..] {
                 "build" => "build-",

--- a/tests/cfg.rs
+++ b/tests/cfg.rs
@@ -200,8 +200,8 @@ fn works_through_the_registry() {
 
     Package::new("foo", "0.1.0").publish();
     Package::new("bar", "0.1.0")
-            .target_dep("foo", "0.1.0", "'cfg(unix)'")
-            .target_dep("foo", "0.1.0", "'cfg(windows)'")
+            .target_dep("foo", "0.1.0", "cfg(unix)")
+            .target_dep("foo", "0.1.0", "cfg(windows)")
             .publish();
 
     let p = project("a")

--- a/tests/registry.rs
+++ b/tests/registry.rs
@@ -1,9 +1,11 @@
 #[macro_use]
 extern crate cargotest;
 extern crate hamcrest;
+extern crate url;
 
 use std::fs::{self, File};
 use std::io::prelude::*;
+use std::path::PathBuf;
 
 use cargotest::cargo_process;
 use cargotest::support::git;
@@ -11,6 +13,10 @@ use cargotest::support::paths::{self, CargoPathExt};
 use cargotest::support::registry::{self, Package};
 use cargotest::support::{project, execs};
 use hamcrest::assert_that;
+use url::Url;
+
+fn registry_path() -> PathBuf { paths::root().join("registry") }
+fn registry() -> Url { Url::from_file_path(&*registry_path()).ok().unwrap() }
 
 #[test]
 fn simple() {
@@ -609,7 +615,9 @@ fn bad_license_file() {
         .file("src/main.rs", r#"
             fn main() {}
         "#);
-    assert_that(p.cargo_process("publish").arg("-v"),
+    assert_that(p.cargo_process("publish")
+                 .arg("-v")
+                 .arg("--host").arg(registry().to_string()),
                 execs().with_status(101)
                        .with_stderr_contains("\
 [ERROR] the license file `foo` does not exist"));
@@ -1339,4 +1347,38 @@ this warning.
 [COMPILING] [..]
 [FINISHED] [..]
 "));
+}
+
+#[test]
+fn toml_lies_but_index_is_truth() {
+    Package::new("foo", "0.2.0").publish();
+    Package::new("bar", "0.3.0")
+            .dep("foo", "0.2.0")
+			.file("Cargo.toml", r#"
+				[project]
+                name = "bar"
+                version = "0.3.0"
+                authors = []
+
+                [dependencies]
+                foo = "0.1.0"
+			"#)
+            .file("src/lib.rs", "extern crate foo;")
+			.publish();
+
+    let p = project("foo")
+        .file("Cargo.toml", r#"
+            [project]
+            name = "bar"
+            version = "0.5.0"
+            authors = []
+
+            [dependencies]
+            bar = "0.3"
+        "#)
+        .file("src/main.rs", "fn main() {}");
+    p.build();
+
+    assert_that(p.cargo("build").arg("-v"),
+                execs().with_status(0));
 }

